src/Control/Distributed/Process/Internal/Primitives.hs

```diff
@@ -917,9 +919,17 @@ unlinkPort sport = do
 unmonitor :: MonitorRef -> Process ()
 unmonitor ref = do
   unmonitorAsync ref
-  receiveWait [ matchIf (\(DidUnmonitor ref') -> ref' == ref)
-                        (\_ -> return ())
-              ]
+  receiveWait [ matchIf isDidUnmonitor ignore ]
+  receiveAllNow [ matchIf isProcessMonitorNotification ignore,
```

Comment:
I guess there is no need to collect a list of units `[()]` as receiveAllNow was defined to do below. How about writing this in a single loop here:
```
fix $ \loop ->
  m <- receiveTimeout 0 [...]
  forM_ m loop
```